### ISSUE TITLE
Not sure I agree with PIE807

### ISSUE THREAD

**Issue by ezyang** [astral-sh/ruff#10141]
When I say `lambda: {}`, it is clear that this callable takes no arguments, and I will raise an error if you pass me arguments. When you transform this into `dict`, now I could potentially take arguments, and I can only rely on an appropriate type signature to prevent people from inappropriately feeding arguments into this lambda.

cc @Skylion007

**Comment by zanieb** [comment]
Could you provide a full example so people know what this looks like in context?

**Comment by ezyang** [comment]
Sure thing:

```
def trace_structured(
    name: str,
    # NB: metadata expected to be dict so adding more info is forward compatible
    # Tuple[str, int] is a special case for string interning
    metadata_fn: Callable[[], Union[Dict[str, Any], Tuple[str, int]]] = dict,
    *,
    payload_fn: Callable[[], Optional[Union[str, object]]] = lambda: None,
    suppress_context: bool = False,
):  
    """ 
    metadata is an arbitrary JSON compatible struct, but it's expected to not be
    too long (e.g., less than 1MB)
        
    payload is an arbitrary string, which can be arbitrarily long (but expected to have
    newlines so no lines are too long)
    """ 
    assert "name" not in ["rank", "frame_id", "frame_compile_id", "attempt"]
    assert callable(
        metadata_fn
    ), f"metadata_fn should be callable, but got {type(metadata_fn)}"
    assert callable(
        payload_fn
    ), f"payload_fn should be callable, but got {type(payload_fn)}"
    if trace_log.isEnabledFor(logging.DEBUG):
        record: Dict[str, object] = {}
        record[name] = metadata_fn()
```

`metadata_fn` default is the relevant spot.

**Comment by MichaReiser** [comment]
Playground for the provided example: [link](https://play.ruff.rs/1c40b272-2e86-4f11-8dd6-067ffa1969ca)

**Comment by mikeleppane** [comment]
Hmm...I don't think it changes the inferred type. `lambda: {}` vs `dict` is TMU basically the same from the type checker's point of view. You can verify this using `reveal_type` (mypy).
In an untyped Python code base you can do whatever you like, until it crashes or not at runtime.

**Comment by AlexWaygood** [comment]
Interestingly, `lambda: {}` seems, if anything, to be slightly faster, even though it's a pure-Python function whereas the `dict` constructor is written in C? I assume this is because the `dict` constructor has to do some pretty complex argument parsing, whereas `lambda: {}` does not.

```sh
% py -m timeit -n 50000000 -s "x = dict" "x()"                            ~/dev
50000000 loops, best of 5: 18.8 nsec per loop
% py -m timeit -n 50000000 -s "x = lambda: {}" "x()"                      ~/dev
50000000 loops, best of 5: 18.2 nsec per loop
```

**Comment by AlexWaygood** [comment]
I still feel like I prefer `dict` and `list` from an aesthetic point of view to `lambda: {}` and `lambda: []`, and these are often options that Python programmers forget. So I do feel like PIE807 provides some value here. The performance difference is absolutely tiny, and I'm guessing the difference in the callback signature is _probably_ only going to matter to you if you're the kind of person who's already using type annotations -- in which case, the transformation is likely safe, since if you want to enforce a narrower signature for the callback function, you can easily use type annotations to do so.

That's my opinion -- but the points raised here were definitely interesting to think about. I can see the argument that `lambda: {}` is more explicit to readers of the code that you should never pass any arguments to the callback function.

**Comment by ezyang** [comment]
As a Haskeller, I don't mind doing eta reduction as long as the expressions actually mean the same thing... which they don't in this case. I am sympathetic to the defaultdict(lambda: []) case though, which I do agree is far more idiomatic as defaultdict(list). When the lambda is a default argument for the function, I feel matters are different, since it's in a contravariant position rather than a covariant position.

**Comment by ThiefMaster** [comment]
I disabled PIE807 in my codebase for a similar reason: `lambda: {}` and `lambda: []` are clear ways of indicating that I'm providing a callable that returns an empty (mutable) object.  In places like Marshmallow schemas (`load_default=lambda: []` etc.) this is much clearer than `load_default=dict` which this rule would want me to use.

**Comment by henryiii** [comment]
It's only faster on Python 3.12; on 3.11:

```console
$ python3.11 -m timeit -n 50000000 -s "x = dict" "x()"
50000000 loops, best of 5: 45.5 nsec per loop
$ python3.11 -m timeit -n 50000000 -s "x = lambda: {}" "x()"
50000000 loops, best of 5: 48.3 nsec per loop
```

On 3.12, it is a bit faster:

```console
$ python3.12 -m timeit -n 50000000 -s "x = dict" "x()"
50000000 loops, best of 5: 49.4 nsec per loop
$ python3.12 -m timeit -n 50000000 -s "x = lambda: {}" "x()"
50000000 loops, best of 5: 44.7 nsec per loop
```

It might be taking advantage of the fact that `{}` is about twice as fast as `dict()` as it avoids the function lookup overhead (like how `f"{x}"` is faster than `str(x)`), it seems 3.12 optimizes on the lambda just a bit better, perhaps. Also benchmarks at this level are pretty tricky.

For fun, I checked 3.13a3 with nogil and JIT enabled, and it seems the dict one has gone up:

```console
$ PYTHON_CONFIGURE_OPTS='--enable-optimizations --with-lto --enable-experimental-jit --disable-gil' PYTHON_CFLAGS='-march=native -mtune=native' pyenv install 3.13.0a3
$ ~/.pyenv/versions/3.13.0a3/bin/python -m timeit -n 50000000 -s "x = dict" "x()"
50000000 loops, best of 5: 62.9 nsec per loop
$ ~/.pyenv/versions/3.13.0a3/bin/python -m timeit -n 50000000 -s "x = lambda: {}" "x()"
50000000 loops, best of 5: 49.1 nsec per loop
```

I like "dict" better (once you get used to functions as objects), just was curious when I saw the perf comparison.

**Comment by zanieb** [comment]
Is the open question here whether we should reduce the scope of `PIE807`? What specific case would we no longer enforce?
